components/service/glean/src/main/java/mozilla/components/service/glean/storages/StringListsStorageEngine.kt

```diff
@@ -0,0 +1,82 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.annotation.SuppressLint
+import mozilla.components.service.glean.CommonMetricData
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This singleton handles the in-memory storage logic for strings. It is meant to be used by
+ * the Specific Strings API and the ping assembling objects. No validation on the stored data
+ * is performed at this point: validation must be performed by the Specific Strings API.
+ *
+ * This class contains a reference to the Android application Context. While the IDE warns
+ * us that this could leak, the application context lives as long as the application and this
+ * object. For this reason, we should be safe to suppress the IDE warning.
+ */
+@SuppressLint("StaticFieldLeak")
+internal object StringListsStorageEngine : StringListsStorageEngineImplementation()
+
+internal open class StringListsStorageEngineImplementation(
+    override val logger: Logger = Logger("glean/StringsStorageEngine")
+) : GenericScalarStorageEngine<Array<String>?>() {
+
+    companion object {
+        // Maximum length of any list
+        private const val MAX_LIST_LENGTH_VALUE = 50
+    }
+
+    @SuppressWarnings("unchecked")
+    override fun deserializeSingleMetric(value: Any?): Array<String>? {
+        return value as? Array<String>
+    }
+
+    /**
+     * Record a string in the desired stores.
+     *
+     * @param metricData object with metric settings
+     * @param value the string value to add
+     */
+    @Synchronized
+    fun add(
+        metricData: CommonMetricData,
+        value: List<String>
+    ) {
+        // Use a custom combiner to add the string to the existing list rather than overwriting
+        super.recordScalar(metricData, value.toTypedArray()) { currentValue, newValue ->
+            currentValue?.let {
+                return@let if (it.count() < MAX_LIST_LENGTH_VALUE) {
+                    it + newValue!!
+                } else {
+                    logger.warn("List [${metricData.category}.${metricData.name}] " +
+                            "length ${it.count()} > $MAX_LIST_LENGTH_VALUE")
+                    it
+                }
+            } ?: newValue
+        }
+    }
+
+    /**
+     * Record a string in the desired stores.
+     *
+     * @param metricData object with metric settings
+     * @param value the string list value to record
+     */
+    @Synchronized
+    fun set(
+        metricData: CommonMetricData,
+        value: MutableList<String>
+    ) {
+        if (value.count() > MAX_LIST_LENGTH_VALUE) {
+            logger.warn("List [${metricData.category}.${metricData.name}] " +
+                    "length ${value.count()} > $MAX_LIST_LENGTH_VALUE")
+            // Truncate list to maximum length
+            super.recordScalar(metricData, value.subList(0, MAX_LIST_LENGTH_VALUE).toTypedArray())
```

Comment:
What about always truncating this instead of having the `if` branch?

Reply:
If you try and truncate a list with subList() that isn't as long as the value you are truncating to, you get an index out of bounds exception.  I'll play around with some ways of trying to minimize the branching here and see what I come up with.

Reply:
Good point! What about using the [take](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/take.html) function? It should work with `value.take(MAX_LIST_LENGTH_VALUE)`. That reminds me... we should also have test coverage to make sure the different cases here are covered:

- trying to record less items than the limit;
- more items than the limit in a single accumulation;
- less items than the limit, but exceeding the limit if counting the currently stored one as well